add cluster state (splits, moves, reboots, creates) test

nginthfs opened the pull request:
This PR adds `splits-moves-reboots.json`, as well as a new `TaskType`: `move-shard`. Currently our tests are focused on indexing and querying, so this test is meant to show differences in performance for state-based operations like collection creation, splits, and moves.

patsonluk commented on src/main/java/StressMain.java:
```diff
@@ -649,6 +655,54 @@ private static Callable taskCallable(Workflow workflow, SolrCloud cloud, Map<Str
+			} else if (type.moveReplica != null) {
+				long taskStart = System.currentTimeMillis();
+
+				String collectionName = params.get("COLLECTION");
```
should we use `type.moveReplica.collection` here instead?

nginthfs replied:
To be honest, the `resolveString` stuff was just super confusing to me, and getting `COLLECTION` from the params just worked. But that should work, and it would give a purpose to `type.moveReplica.collection`, so I'll do that!

patsonluk commented on src/main/java/StressMain.java:
```diff
@@ -649,6 +655,54 @@ private static Callable taskCallable(Workflow workflow, SolrCloud cloud, Map<Str
+				Collections.shuffle(copyOfSolrNodes);
+				for (SolrNode sn : copyOfSolrNodes) {
+					String toNodeName = sn.getNodeName() + "_solr";
+					if (toNodeName != fromNodeName) {
```
minor nit - little bit safer to use string.equals

nginthfs replied:
Ah of course, bit rusty on my Java skills

patsonluk commented on src/main/java/StressMain.java:
```diff
@@ -649,6 +655,54 @@ private static Callable taskCallable(Workflow workflow, SolrCloud cloud, Map<Str
+				}
+				log.info(String.format("moving replica %s from %s to %s", replicaName, fromNodeName, targetNodeName));
+
+				try (HttpSolrClient client = new HttpSolrClient.Builder(cloud.nodes.get(0).getBaseUrl()).build();) {
```
Be nice if we can discover the os node here

patsonluk commented on src/main/java/StressMain.java:
```diff
@@ -649,6 +655,54 @@ private static Callable taskCallable(Workflow workflow, SolrCloud cloud, Map<Str
+					finalResults.get(taskName).add(Map.of("total-time", totalTime, "start-time", (taskStart- executionStart)/1000.0,
+							"end-time", (taskEnd- executionStart)/1000.0,
+							"init-timestamp", executionStart, "start-timestamp", taskStart, "end-timestamp", taskEnd));
+				} catch (Exception ex) {
```
Let's not catch the exception here and let it bubble up. I know some of the older parts of the code logic still catch the exception but newer ones should just throw exception as is.

This allow the executor to properly detect the exception and skip all the follow tasks that depending on a failing tasks :)

nginthfs replied:
Oops, yep, I agree. I think I was doing some debugging here and just left this in.

patsonluk commented:
Thanks for tackling this @nginthfs . The move replica (or moveshard in solrman) is actually a pretty tricky problem!

In solrman, we are not using `movereplica`, instead we use `addreplica` then `deletereplica`. (and also we call that moveshard instead of movereplica, which i also found a bit confusing in the solrman code). Not 100% sure why we didn't use `movereplica` (it seems to exist back in solr 6), but in order to test our current flow, we might want to test `addreplica` and `deletereplica`.

This however, leads to some deeper questions/issues:
1. If we implement such `movereplica` test as `addreplica` and `deletereplica`, then we are adding something very specific to FullStory, which does not make too much sense to solr-bench itself.
2. I do like your shard-splitting and collection creation tests which uses direct URI - as this is what we use in solrman. Is it possible to just test `addreplica` and `deletereplica` similarly w/o needing a specific `move-replica` task? From the look of it, im pretty sure you have considered that before going down this current path 😓 , and i'm wondering what was the blocker? was it the difficult to figure out which node did a replica land after collection creation? Can we use flags such as `createNodeSet` (https://solr.apache.org/guide/8_8/collection-management.html#collection-management) to tackle it? 🤔

nginthfs commented on src/main/java/StressMain.java:
```diff
@@ -649,6 +655,54 @@ private static Callable taskCallable(Workflow workflow, SolrCloud cloud, Map<Str
+
+				String collectionName = params.get("COLLECTION");
+				String replicaName = "", fromNodeName = "";
+				try (CloudSolrClient client = new CloudSolrClient.Builder(List.of(cloud.getZookeeperUrl()), Optional.ofNullable(cloud.getZookeeperChroot())).build();) {
```
Ah, wasn't aware of that function, I'll change that!

nginthfs commented:
> Thanks for tackling this @nginthfs . The move replica (or moveshard in solrman) is actually a pretty tricky problem!
> 
> In solrman, we are not using `movereplica`, instead we use `addreplica` then `deletereplica`. (and also we call that moveshard instead of movereplica, which i also found a bit confusing in the solrman code). Not 100% sure why we didn't use `movereplica` (it seems to exist back in solr 6), but in order to test our current flow, we might want to test `addreplica` and `deletereplica`.
> 
> This however, leads to some deeper questions/issues:
> 
> 1. If we implement such `movereplica` test as `addreplica` and `deletereplica`, then we are adding something very specific to FullStory, which does not make too much sense to solr-bench itself.
> 2. I do like your shard-splitting and collection creation tests which uses direct command/URI - as this is what we use in solrman. Is it possible to just test `addreplica` and `deletereplica` similarly w/o needing a specific `move-replica` task? From the look of it, im pretty sure you have considered that before going down this current path 😓 , and i'm wondering what was the blocker? was it difficult to figure out which node did a replica land after collection creation? Can we use flags such as `createNodeSet` (https://solr.apache.org/guide/8_8/collection-management.html#collection-management) to tackle it? 🤔

Ah, using `addreplica` and `deletereplica` actually makes this way simpler. I can get the shard without needing a custom task (`moveReplica`). And then yeah, I could go ahead and use `createNodeSet` to randomly place replicas across all the nodes. I'll implement it this way for this test, but since I've done the `movereplica` task work, might as well leave that in? That had to be a custom task because for `MOVEREPLICA` you need to know the target node and the source node, and I couldn't see a simple way of doing that with parameter.

nginthfs commented:
Alright @patsonluk I think I've addressed all your comments, thank you! For the `MOVEREPLICA` vs. `ADDREPLICA`/`DELETEREPLICA` part, I've added a `Boolean` argument to `TaskType.MoveReplica` called `useAddDelete`, which instructs the task to use add/delete replica like we do in Solrman. It defaults to using `MOVEREPLICA`.

patsonluk commented on src/main/java/StressMain.java:
```diff
@@ -690,6 +749,19 @@ private static CloudSolrClient buildSolrClient(SolrCloud cloud) {
+	 * buildHttpSolrClient tries to build the client pointing to the overseer node, otherwise
+	 * it returns a client pointing to the first node in cloud.nodes
+	 */
+	private static HttpSolrClient buildHttpSolrClient(SolrCloud cloud, String overseerLeader) {
```
Nice to have this!! Should we name overseerLeader even more generic such as preferredNodeName? I'm thinking this could be useful for other usages as well (preferred node to send query to or index data to?)

patsonluk commented:
Sorry about the late reply! 😓 

For sure, let's keep the `MoveReplica` task type you implemented as it could be useful for others! (that does use MOVEREPLICA command)

As for the add and delete replica, im wondering if we can control it only from the config? Since using Add and delete replica seem really specific to our own usage (again, I haven't checked the history - why we didn't use movereplica at first place 😅 ), and if we can test our special usage purely from config, that would be cleaner 😊 . Something along the line of

```
"add-replica": {
      "command": "${SOLRURL}admin/collections?action=ADDREPLICA&name=collection${INDEX}&shard=shard1",
      "defaults": {
        "INDEX": 0
      }
    },
"delete-replica":{
    "command": "${SOLRURL}admin/collections?action=DELETEREPLICA&name=collection${INDEX}&shard=shard1&replica=collection${INDEX}_shard1_replica2",
      "defaults": {
        "INDEX": 0
      }
}    
```

Not sure if the above would work since I might have overlooked some details! But if it works that we can skip adding extra code logic to emulate our odd usage 😁 


The rest of the code looks really good! Thanks for tackling this again! 💪🏼

nginthfs commented:
I considered doing it like that but I was concerned about the replica name for `DELETEREPLICA`. I'm still unfamiliar with Solr a bit, are the replica names always consistent where the first replica will be of this form `collection${INDEX}_shard1_replica2`? I'm guessing that's so, since you recommended it 😅

patsonluk commented:
> I considered doing it like that but I was concerned about the replica name for `DELETEREPLICA`. I'm still unfamiliar with Solr a bit, are the replica names always consistent where the first replica will be of this form `collection${INDEX}_shard1_replica2`? I'm guessing that's so, since you recommended it 😅

😁 Sorry I was just making the suggestion by just guessing that it would work, but by looking at the code a bit deeper, it seems non-deterministic for non PRS collection and deterministic for PRS collection. 

Want to share what I found here so we could share the knowledge (and to my future self 😅 ):
- The replica itself has (at least) 2 names, by inspecting the class `Replica` and the code logic that built it:
  - The `name` field in `Replica` is something called the `coreNodeName` in the invoking logic. It has a shorter form of `core_node<n1>`, and is the id used for Replica Deletion, however it might not be as visible otherwise as most responses show "core name" (below) instead of replica name
  - The `coreName` field in `Replica` is also another distinct name we can find in `Replica`, core name is in the format of `<collection>_<shard>_replica_n<n2>`

Now both "core name" and "core node name" are assigned on creation reading and increment on the "counter" value stored in zk path `solr/collections/<collection>/counter`, which is the `n1` and `n2` above. Take note that they are not the same values. (but usually pretty close)

For non PRS collection creation, it's non-deterministic as:
1. `CREATECOLLECTION` handling in overseer iterates on the `numShards` defined to create each core, assign `n` in the core name
2. Send a `ADDREPLICA` command call to itself (the overseer), which assign another `n` to the "core node name" (the `name` field in `Replica`)
3. Now point 1 and point 2 can be executed concurrently therefore the `n` could look different from run to run

For PRS collection creation, in point 2, it no longer sends the `ADDREPLICA` command to itself but instead execute it within the same execution flow/thread, therefore is supposed to be deterministic for new collection creation, but things could get weird once we split/move shards.

Now back to the actual question...can we control/predict the replica name (core node name) that is generated from `ADDREPLICA`. 

As for controlling, there's a message prop "core_node_name" that can override the name generation, but this is only available to internal calls within Solr, from API no param can control it (there's `name` which will control the core name but not the core node name)

As for predicting, this one is a bit more interesting. Even though the initial replicas/cores in collection creation are not too deterministic, the extra replica add (assuming no extra operations that fiddle with `solr/collections/<collection>/counter` after collection creation) is supposed to have core name `<collection>_<shard>_replica_n<2*x + 1>` and core node name `core_node<2*x + 2>`, which x is the total # of replicas created so far before the ADDREPLICA command. Therefore, if we created a new collection with 8 shards, and then add an extra replica, the replica name (core node name) is supposed to be `core_node18`. I tested several times locally and that appears to be true.

@nginthfs do u think it's possible to try the above theory and see if it works? 😊 





![image](https://user-images.githubusercontent.com/2895902/232113410-a90d924d-3ffa-4fa8-a35c-791f9b90159e.png)

patsonluk commented:
> First of all, thank you so much for the investigation on this, it's nice to know how the replica creation works! I've tested this locally and I think I've confirmed the same thing you discovered. My concern here is that this would be tied to implementation, and if how the counter worked or how replica names are assigned changed in the future for some reason, then this test would break. However, I think I may have a solution here: since we don't necessarily care _which_ replica gets deleted (for the test's sake), we could use the `count` argument for `DELETEREPLICA`. So I could do
> 
> ```
> "add-replica": {
>       "command": "${SOLRURL}admin/collections?action=ADDREPLICA&name=collection${INDEX}&shard=shard1",
>       "defaults": {
>         "INDEX": 0
>       }
>     },
> "delete-replica":{
>     "command": "${SOLRURL}admin/collections?action=DELETEREPLICA&name=collection${INDEX}&shard=shard1&count=1",
>       "defaults": {
>         "INDEX": 0
>       }
> }    
> ```
> 
> I believe this should be testing almost the same amount of work that we do right now in Solrman to move a replica, but let me know what you think!

Totally! It's true that we don't really care that much on which one gets deleted as far as we test deletion! 😊

patsonluk reviewed (approved):
LGTM! 💪🏼
